frame/beefy-mmr/primitives/src/lib.rs

```diff
@@ -25,84 +25,45 @@
 //!
 //! Merkle Tree is constructed from arbitrary-length leaves, that are initially hashed using the
-//! same [Hasher] as the inner nodes.
+//! same hasher as the inner nodes.
 //! Inner nodes are created by concatenating child hashes and hashing again. The implementation
 //! does not perform any sorting of the input data (leaves) nor when inner nodes are created.
 //!
 //! If the number of leaves is not even, last leave (hash of) is promoted to the upper layer.
 
-#[cfg(not(feature = "std"))]
-extern crate alloc;
-#[cfg(not(feature = "std"))]
-use alloc::vec::Vec;
+pub use sp_runtime::traits::Keccak256;
+use sp_runtime::{app_crypto::sp_core, sp_std, traits::Hash as HashT};
+use sp_std::{vec, vec::Vec};
 
 use beefy_primitives::mmr::{BeefyAuthoritySet, BeefyNextAuthoritySet};
 
-/// Supported hashing output size.
-///
-/// The size is restricted to 32 bytes to allow for a more optimised implementation.
-pub type Hash = [u8; 32];
-
-/// Generic hasher trait.
-///
-/// Implement the function to support custom way of hashing data.
-/// The implementation must return a [Hash](type@Hash) type, so only 32-byte output hashes are
-/// supported.
-pub trait Hasher {
-	/// Hash given arbitrary-length piece of data.
-	fn hash(data: &[u8]) -> Hash;
-}
-
-#[cfg(feature = "keccak")]
-mod keccak256 {
-	use tiny_keccak::{Hasher as _, Keccak};
-
-	/// Keccak256 hasher implementation.
-	pub struct Keccak256;
-	impl Keccak256 {
-		/// Hash given data.
-		pub fn hash(data: &[u8]) -> super::Hash {
-			<Keccak256 as super::Hasher>::hash(data)
-		}
-	}
-	impl super::Hasher for Keccak256 {
-		fn hash(data: &[u8]) -> super::Hash {
-			let mut keccak = Keccak::v256();
-			keccak.update(data);
-			let mut output = [0_u8; 32];
-			keccak.finalize(&mut output);
-			output
-		}
-	}
-}
-#[cfg(feature = "keccak")]
-pub use keccak256::Keccak256;
-
 /// Construct a root hash of a Binary Merkle Tree created from given leaves.
 ///
 /// See crate-level docs for details about Merkle Tree construction.
 ///
 /// In case an empty list of leaves is passed the function returns a 0-filled hash.
-pub fn merkle_root<H, I, T>(leaves: I) -> Hash
+pub fn merkle_root<H, I>(leaves: I) -> H::Output
 where
-	H: Hasher,
-	I: IntoIterator<Item = T>,
-	T: AsRef<[u8]>,
+	H: HashT,
+	H::Output: Default + AsRef<[u8]>,
+	I: IntoIterator,
+	I::Item: AsRef<[u8]>,
 {
-	let iter = leaves.into_iter().map(|l| H::hash(l.as_ref()));
-	merkelize::<H, _, _>(iter, &mut ())
+	let iter = leaves.into_iter().map(|l| <H as HashT>::hash(l.as_ref()));
+	merkelize::<H, _, _>(iter, &mut ()).into()
 }
 
-fn merkelize<H, V, I>(leaves: I, visitor: &mut V) -> Hash
+fn merkelize<H, V, I>(leaves: I, visitor: &mut V) -> H::Output
 where
-	H: Hasher,
-	V: Visitor,
-	I: Iterator<Item = Hash>,
+	H: HashT,
+	H::Output: Default + AsRef<[u8]>,
+	V: Visitor<H::Output>,
+	I: Iterator<Item = H::Output>,
 {
-	let upper = Vec::with_capacity(leaves.size_hint().0);
+	let upper = Vec::with_capacity(leaves.size_hint().1.unwrap());
```

Comment:
Could you, please, change `.unwrap()` to `.expect("<why-you-think-noone-will-ever-see-this-text; qed>")` (see https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md#style for details).

Reply:
Good catch! I would just change it to
```suggestion
	let upper = Vec::with_capacity(leaves.size_hint().1.unwrap_or(0));
```

Reply:
Sorry, this was just an experiment. Normally for `Vec` `size_hint().1` will always return `Some(len)`, but probably it doesn't make sense for other type of iterators. I rolled it back.

Reply:
Hmmm `unwrap_or(0)` also sounds good.

Reply:
Done. Changed it to `leaves.size_hint().1.unwrap_or(0).saturating_add(1)) / 2`